src/hotspot/share/nmt/virtualMemoryTrackerWithTree.cpp

```diff
@@ -0,0 +1,198 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "logging/log.hpp"
+#include "nmt/virtualMemoryTrackerWithTree.hpp"
+#include "nmt/memTracker.hpp"
+
+RegionsTree* VirtualMemoryTrackerWithTree::_tree;
+
+bool VirtualMemoryTrackerWithTree::initialize(NMT_TrackingLevel level) {
+  assert(_tree == nullptr, "only call once");
+  if (level >= NMT_summary) {
+    _tree = new (std::nothrow, mtNMT) RegionsTree(level == NMT_detail);
+    return (_tree != nullptr);
+  }
+  return true;
+}
+
+
+bool VirtualMemoryTrackerWithTree::add_reserved_region(address base_addr, size_t size,
+  const NativeCallStack& stack, MEMFLAGS flag) {
+  assert(_tree != nullptr, "Sanity check");
+  if (flag == mtTest) {
+    log_debug(nmt)("add reserve rgn, base: " INTPTR_FORMAT " end: " INTPTR_FORMAT, p2i(base_addr), p2i(base_addr + size));
+  }
+  VMATree::SummaryDiff diff = _tree->reserve_mapping((size_t)base_addr, size, _tree->make_region_data(stack, flag));
+  apply_summary_diff(diff);
+  return true;
+
+}
+
+void VirtualMemoryTrackerWithTree::set_reserved_region_type(address addr, MEMFLAGS flag) {
+    ReservedMemoryRegion rgn;
+    _tree->find_reserved_region(addr, &rgn);
+    if (rgn.flag() == flag)
+      return;
+
+    const VMATree::position& start = (VMATree::position)addr;
+    const VMATree::position& end = (VMATree::position)(rgn.end() + 1);
+    RegionsTree::NodeHelper* prev = nullptr;
+    if (start > end) {
+      _tree->dump(tty);
+      tty->print_cr("requested addr: " INTPTR_FORMAT " end: " INTPTR_FORMAT, p2i(addr), p2i((address) end));
+      _tree->find_reserved_region(addr, &rgn, true);
+    }
+    size_t rgn_size = 0;
+    size_t comm_size = 0;
+    MEMFLAGS old_flag = mtNone, bak_out_flag;
+    bool base_flag_set = false;
+    bool release_node_found = false;
+    _tree->visit_range_in_order(start, end, [&](VMATree::TreapNode* node){
+      RegionsTree::NodeHelper* curr = (RegionsTree::NodeHelper*)node;
+      if (!base_flag_set) {
+        old_flag = curr->out_flag();
+        base_flag_set = true;
+      }
+      bak_out_flag = curr->out_flag();
+      curr->set_out_flag(flag);
+      if (prev != nullptr) {
+        curr->set_in_flag(flag);
+        rgn_size += curr->distance_from(prev);
+        if (prev->is_committed_begin())
+          comm_size += curr->distance_from(prev);
+      }
+      prev = curr;
+      if (curr->is_released_begin() || bak_out_flag != old_flag) {
+        if (bak_out_flag != old_flag) {
+          curr->set_out_flag(bak_out_flag);
+        }
+        VirtualMemorySummary::move_reserved_memory(old_flag, flag, rgn_size);
+        VirtualMemorySummary::move_committed_memory(old_flag, flag, comm_size);
+        release_node_found = true;
+        return false;
+      }
+      return true;
+    });
+}
+
+void VirtualMemoryTrackerWithTree::apply_summary_diff(VMATree::SummaryDiff diff) {
```

Comment:
Applying a summary diff in the MemoryFileTracker is this:

```c++
  for (int i = 0; i < mt_number_of_types; i++) {
    VirtualMemory* summary = file->_summary.by_type(NMTUtil::index_to_flag(i));
    summary->reserve_memory(diff.flag[i].commit);
    summary->commit_memory(diff.flag[i].commit);
  }
```

This is much simpler and doesn't require looking at signs and so on.

Reply:
In `MemoryFileTracker`, the changes in commit/reserve are applied to a local `VirtualMemorySnapshot`. Here we have to apply them to the global `VirtualMemorySummary`.

Reply:
The main purpose of the `if (...)` cases is to find if the request to apply the delta is valid or not. There are related assertions in VirtualMemory but not so informative. Also, using `log_debug` lets the build proceed and just show the messages.
These messages help a lot when something goes wrong in terms of commit/uncommit/release failure.

Reply:
How does my example code not account for this? You can still get rid of the sign checking.

Reply:
Case of 'commit' error is missing from your suggestion. At commit time, ` c > reserved` is invalid too.
`(size_t)-r` for positive `r` is a large number and is greater than `reserved`. 
We have to find out the intent of the call first (by checking the sign of `r` or `c`) and then check if it is valid or not.